vxsort: Add Arm64 Neon implementation and tests

a74nh opened the pull request:
Add an implementation of vxsort for Neon.

Add a testing framework to be able to build vxsort by itself, run some basic sanity tests, run some basic performance tests. This will be useful should further improvements be made, or other targets (SVE?) added.

a74nh committed:
vxsort: Add Arm64 Neon implementation and tests

a74nh commented:
Work in progress.

Currently the testing for vxsort exists in `src/coreclr/gc/vxsort/standalone`. This needs refactoring and moving into `src/tests` somewhere.

I still need to add bitonic search and packing support for Neon. The searching for small lists currently uses a copy of insertsort instead of bitonic search. So that I can check performance, for AVX2 I've disabled packing and switch to insertsort too.

Performance testing is very basic, but running `./simple_bench/Project_demo 250` on Cobalt 100 I see roughly the same for both vxsort and insertsort:
```
vxsort: Time= 3 us
vxsort: Time= 5 us
vxsort: Time= 4 us
insertsort: Time= 5 us
insertsort: Time= 3 us
insertsort: Time= 7 us
```

On an AVX2 X64 (Gold 5120T), the vxsort is _slightly_ faster.
```
vxsort: Time= 6 us
vxsort: Time= 6 us
vxsort: Time= 6 us
insertsort: Time= 8 us
insertsort: Time= 6 us
insertsort: Time= 5 us
```

On the same AVX2 X64 (Gold 5120T), switching the vxsort code to use bitonic search and packing:
```
vxsort: Time= 3 us
vxsort: Time= 5 us
vxsort: Time= 4 us
```

Given the above, I'm fairly confident that implementing the rest for Neon with give some improvements.

I noticed that for more than 255 the program segfaults on both X64 and Arm64. This looks like a limitation of vxsort. Might be worth adding some asserts to check the size of the list?

a74nh commented:
@kunalspathak @JulieLeeMSFT

Maoni0 commented:
thanks for your interest in this!

@damageboy has many tests in his repo - https://github.com/damageboy/vxsort-cpp

> I noticed that for more than 255 the program segfaults on both X64 and Arm64.

is 255 number of elements in the array? that'd be quite surprising because we don't even start invoking vxsort till we have at least 8k for avx2 and 128k for avx512.

am11 commented on src/coreclr/gc/vxsort/do_vxsort_neon.cpp:
```diff
@@ -0,0 +1,19 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#if defined(TARGET_ARM64)
```
Maybe just conditionally compile them (in cmake) instead of adding `ifdef`s?

a74nh replied:
> Maybe just conditionally compile them (condition in cmake) instead of adding `ifdef`s?

I'm used to seeing files elsewhere in coreclr be compiled for all targets and just have an ifdef at the top. Eg, anything in jit/ with a cpu in the filename. So, was following that convention. But happy to switch.

am11 commented on src/coreclr/gc/vxsort/introsort.cpp:
```diff
@@ -0,0 +1,108 @@
+#include "introsort.h"
+
+
+ void introsort::sort (uint8_t** begin, uint8_t** end, int ignored)
+    {
```
nit: formatting / extra spaces
```suggestion
void introsort::sort (uint8_t** begin, uint8_t** end, int ignored)
{
```

a74nh committed:
allow more than 255 elements

a74nh committed:
cleanup insertsort

kunalspathak commented:
Thanks @a74nh . Can you confirm which of the above numbers are with vs. without your change on Cobalt 100?

kunalspathak commented:
> Thanks @a74nh . Can you confirm which of the above numbers are with vs. without your change on Cobalt 100?

ignore...so seems today we use `insertsort` on arm64, so with your numbers, seems like 30% improvement.

a74nh commented:
> > Thanks @a74nh . Can you confirm which of the above numbers are with vs. without your change on Cobalt 100?
> 
> ignore...so seems today we use `insertsort` on arm64, so with your numbers, seems like 30% improvement.

Yes. I'm hoping to get more by porting both bitonic search and packing for Arm64. In the above figures, I've disabled both on those on X86. When I re-enable them again, X86 goes from ~576ms to ~162ms. So there's definitely some more performance to find.

a74nh committed:
Add Neon to bitonic

a74nh committed:
Add scalar version of bitonic

Change-Id: I19e0fc293b67e28d1dd5491efd9b4e9b86c5c4d7

a74nh committed:
remove Neon from bitonic

a74nh commented:
I've added an implementation of Bitonic.

The plan was to do the bitonic sort using NEON. Unfortunately instructions like `rev`, `min`, `max` etc do not have variants that work on 64bit elements - they only have 8/16/32 variants. (A broken version showing what it would look like if those instructions existed [is here](https://github.com/dotnet/runtime/blob/9c8151d581ce67776e09cdd244f0c4ddf4a33222/src/coreclr/gc/vxsort/smallsort/bitonic_sort.NEON.uint64_t.generated.h)).

For some of the bitonic functions, they can be done in NEON with a few extra instructions (eg `cmgt`+`bit` instead of `max`). For other functions the most optimal way is to move the values into GPR registers and use scalar code. That's very messy and looses perf in all the moves.

An alternative is to simply to avoid NEON and use GPR registers throughout. This can be done by simply writing the code in C++ instead of intrinsics, allowing the compiler to optimise.

As a result, I've implemented the bitonic using scalar code. It's highly doubtful that a mix of NEON+scalar would give better performance. As a bonus it is architecture independent code.

Note that for 8/16/32 values, NEON would be the preferred option. Also, SVE would give better performance on 256bit machines (currently only neoverse V1), but it's doubtful on 128bit machines, although it would shorten the code size. I don't plan on implementing with SVE in this PR.

Trying this code on cobalt 100 shows quite an additional speedup (previously vxsort was running at ~150ms)
```
❯ ./simple_bench/Project_demo 8000
vxsort: Time= 113 us
vxsort: Time= 123 us
vxsort: Time= 117 us
insertsort: Time= 220 us
insertsort: Time= 221 us
insertsort: Time= 238 us
```

In the new year, I'll look at cleaning this up, sorting out the tests etc. I'll also looking at the missing "packing" code in vxsort, see if there's anything else to gain.

a74nh committed:
Remove TARGET_ checks

a74nh committed:
Enable vxsort in GC for Arm64

a74nh committed:
Remove all SVE references

a74nh committed:
Fix nativeaot build

a74nh committed:
Fix ISA detection for windows

a74nh committed:
Add vxsort to windows Arm64 build

a74nh committed:
Merge main

a74nh committed:
Add vxsort to nativeaot

a74nh committed:
Explict popcount implementation for msvc

a74nh committed:
Remove reinterpret_cast for msvc

a74nh committed:
Fix mask_popcount for msvc

a74nh committed:
Fix negative popcount for msvc

a74nh committed:
Add vxsort for Arm64 in BuildIntegration

a74nh committed:
Fix windows nativeaot build

a74nh committed:
Fix linux nativeaot build

a74nh committed:
Add vxsort disabled to Arm64 reproNative

a74nh committed:
Fix up TODOs

a74nh committed:
Fix up M4_SIZE for Amd64

a74nh committed:
Remove standalone testing

a74nh committed:
Restore packing for AVX2

a74nh committed:
Remove extra code in machine_traits.h

a74nh committed:
Remove extra code in machine_traits.h

a74nh committed:
Remove unused vxsort_machine_traits<uint32_t, NEON>

a74nh committed:
Add vxsort to gc/sample for Windows Arm

a74nh committed:
Merge main

Change-Id: I9f8ed75abc05fce0b6fb6b3b4d8895fd62fcb2f5

kunalspathak commented:
/azp run runtime-coreclr gcstress-extra

kunalspathak commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

Maoni0 commented:
have you run the many tests that @damageboy has in his repo? I'll see if he could also take a look. I can't get to this till at least a couple of weeks out since I have a bunch of other urgent things to do (and had no idea that this was coming till I saw the PR). 

presumably you've run this with `_DEBUG` which validates the results against `introsort::sort`? 

I would run with the GC's own stress (the gcstressm mentioned above is not actually stressing the GC) which is documented [here](https://github.com/dotnet/runtime/blob/9cf8a80db3354173cfd25dea4660632861dee0eb/docs/workflow/testing/coreclr/gc-stress-run-readme.md).

MichalStrehovsky reviewed (commented):
The changes under nativeaot and tools/aot look good to me but that's not the main part of this so I'm only leaving as a comment instead of approval.

a74nh commented:
> have you run the many tests that @damageboy has in his repo? I'll see if he could also take a look. I can't get to this till at least a couple of weeks out since I have a bunch of other urgent things to do (and had no idea that this was coming till I saw the PR).
> 
> presumably you've run this with `_DEBUG` which validates the results against `introsort::sort`?
> 
> I would run with the GC's own stress (the gcstressm mentioned above is not actually stressing the GC) which is documented [here](https://github.com/dotnet/runtime/blob/9cf8a80db3354173cfd25dea4660632861dee0eb/docs/workflow/testing/coreclr/gc-stress-run-readme.md).

I have tested with my own hand written tests, which I did include in this PR, but I removed because there was no obvious place to put them. AIUI, everything in src/tests is C# driven. Whereas this testing would need direct access to the vxsort routines.

I've not tried the above suggestions, I'll give them all a go.

a74nh committed:
Revert "Remove standalone testing"

This reverts commit 23950b1e6e4b4581fcc8f9293c7c4bee72847e13.

a74nh commented:
> have you run the many tests that @damageboy has in his repo? 

That requires some work. Damageboy's repo has moved on since the coreclr copy was taken. I tried copying the entire vxsort source directory wholesale back into the Damageboy repo and building from there, but there's a lot of stuff that still needs fixing up (moved headers, different makefiles etc etc). The tests themselves are all named explicitly for the target architecture, so there's then a bunch of new test files that'll need adding for neon. And of course this will all be lost as it's not going into a PR anywhere.

Alternatively, I could try moving the tests into coreclr. They'll still need expanding to create NEON versions though. It'd be nice if they could be checked into coreclr somehow.

> I'll see if he could also take a look.

This would be great. He might eventually want to port it back to his repo.

a74nh committed:
Merge main

a74nh committed:
Merge main

mangod9 commented:
Thanks for the perf data, the numbers look good. Looks like there is more work required to get NativeAOT working? 

Interestingly @MichalStrehovsky recently disabled vxsort by default in NativeAOT for x64 due to size concerns. Would be good to check the size overhead of adding Neon implementation for NativeAOT. 

Also now that main should be 11 soon we can start reviewing and getting this in early for 11

a74nh commented:
> Thanks for the perf data, the numbers look good. Looks like there is more work required to get NativeAOT working?

I thought it was working in an older version. Hopefully it's just a quick fix. Away for a few weeks now, but will take a look when I get back.

> Interestingly @MichalStrehovsky recently disabled vxsort by default in NativeAOT for x64 due to size concerns. Would be good to check the size overhead of adding Neon implementation for NativeAOT.

Will do.

> Also now that main should be 11 soon we can start reviewing and getting this in early for 11

Agreed.

There's also some further work that could be done in a follow on PR - supporting 32bit values, and some other bits. If this PR is showing good improvement, then maybe it's worth getting those done soon too.

a74nh committed:
Merge main

a74nh committed:
Update comments

a74nh committed:
Add 32bit compressed sort

JulieLeeMSFT commented:
CC @VSadov for review.

a74nh committed:
Add N to AVX machine traits

a74nh committed:
Fixup makefile

a74nh committed:
fix constants for msvc

a74nh committed:
better vector constants

a74nh committed:
update makefile

a74nh committed:
Add neon to dummy file

a74nh committed:
Add bitonic_sort.NEON.uint32_t

a74nh committed:
Fix mask for msvc

a74nh committed:
fix idx and maxv for msvc

a74nh committed:
Implement merge sorter

a74nh committed:
Add small sorting type to machine traits

VSadov commented:
> What I'm not sure how to do is to build with my CoreCLR libraries.
I've tried dotnet.sh, but that ends with everything coming from .dotnet/ during the build.

There is a way to use locally built packages with the SDK. 
https://github.com/dotnet/runtime/blob/main/docs/workflow/testing/using-dev-shipping-packages.md

Last time I tried I was able to build either Core or AOT apps with that approach.
(anything that says `10` in the instructions should probably mean `11` now)

VSadov commented:
I looked through the implementation and the reported perf/size diffs. This looks at least as good as x64 implementation. 
The self-consistency checks in chk/dbg should catch invalid sort results.

Since we are at the very beginning of net11, I think the best course of action from here is to merge this and see what happens.

VSadov reviewed (approved):
LGTM. Thanks!!!

jkotas reviewed (approved):
LGTM
